src/browser-service.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2020-present, Elastic NV
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+import { URL } from 'url';
+import { createServer } from 'http';
+import { createProxyServer } from 'http-proxy';
+import { chromium } from 'playwright-chromium';
+
+const proxy = createProxyServer();
+const proxyServer = createServer();
+
+proxyServer.on('upgrade', async function (req, socket, head) {
+  const browserServer = await chromium.launchServer({ headless: true });
+  const wsEndpoint = browserServer.wsEndpoint();
+  const { pathname, origin } = new URL(wsEndpoint);
+  req.url = pathname;
+  console.log(`New browser: ${wsEndpoint}`);
+  proxy.ws(req, socket, head, { target: origin });
```

Comment:
why not pass the full url object? Thought proxy lib supports that option.

Reply:
I think I tried that but didn't work, but we'll have another look for the final version.

Reply:
Yeah ++